airflow/models/taskinstance.py

```diff
@@ -2741,6 +2749,14 @@ def tg2(inp):
         map_index_start = ancestor_map_index * further_count
         return range(map_index_start, map_index_start + further_count)
 
+    def inject_xcoms(self, xcoms: list[dict[str, Any]]):
+        """
+        To inject upstream dependencies' output instead of trying
+        to read from DB when testing a task individually.
+        """
```

Comment:
To be honest, I'm not quite sure about the approach, so would ask for feedback. A potential risk and questions around the current one:

1. The users of the framework may start to use it as a public API, but we may want to change how it works at some point in the future. Is it a valid concern or not?

2. Can the users use `XCom` directly without the API's exposed by `TaskInstance` or the models aren't available in the DAGs, so we cover the scenarios of the interactions with `XCom` on the `TaskInstance` layer? 

--

Alternative options, which I thought of:

1. Pass `xcoms` down the call stack, like to `ti.run(injected_xcoms=xcoms)`. Extending the `ti.run` signature didn't  look quite attractive to me, though may be more clear and explicit in comparison to preliminary "injection". But injection is similar to what we do with `params`: https://github.com/apache/airflow/blob/4e3b5ae7248c2327864f64b25dc7a5bd7705430c/airflow/cli/commands/task_command.py#L590

2. Have a module with a global variable, which we could initialise in `task_command` and then read from either in `TaskInstance.xcom_pull` or `BaseXCom.get_one` / `.get_many`. Global state..

3. Monkeypatch `ti.xcom_pull` (dirty?)

4. Implement a custom `XCom` backend to be used in `xcom.py`, which would read the injected data from memory. 

5. If we are able to run everything inside a common transaction, which is rolled-back at the end of the command execution, then we can just "put" the corresponding XComs in "DB" before running the task?

--

Since I wasn't sure about feasibility and validity of some of the approaches above (especially 4 and 5), I implemented the most straightforward way, similar to what we already have (the `params` reference above), seeking for feedback.

Reply:
> Can the users use XCom directly without the API

They can, via `XCom.get_one` or `get_many`. So I feel the only fully reliable way is to actually write the data to the database so they are available directly from the XCom interface.

We already do something similar for DagRun (if you specify a DagRun that does not actually exist, we would create a temporary one in the database for the `test` run and delete it afterwards), so I can think of a couple of approaches:

1. Write the custom XCom data to a separate table, and read them on `XCom`. Delete them after the test run.
2. Write the custom data to the real XCom table, and delete them afterwards. Emit an error (and refuse to execute) if there is already existing data matching the custom input in the table. This feels reasonable since it would only be possible if you test a task _in an existing DAG run but don’t want to use the actual upstream XComs_, which feels very niche and not a necessary feature to me.

Reply:
> Write the custom data to the real XCom table, and delete them afterwards. Emit an error (and refuse to execute) if there is already existing data matching the custom input in the table. This feels reasonable since it would only be possible if you test a task in an existing DAG run but don’t want to use the actual upstream XComs, which feels very niche and not a necessary feature to me.

For this one I have a concern that in a general case script may exit in a tricky way, even skipping `finally` (https://stackoverflow.com/questions/49262379/does-finally-always-execute-in-python). When we're dealing with a temporary DAG run, then that's relatively fine and is visible from the run id:

https://github.com/apache/airflow/blob/main/airflow/cli/commands/task_command.py#L74-L80

But for a normal DAG run it may leave the data in a strange state. Also probably if this is an ongoing DAG run, then, if I'm not mistaken, there might be race conditions like "write XCom in test command" -> "write the same XCom in a real task executed in the DAG run" -> "delete XCom in test command (clean up)".

What do you think?